### PR TITLE
[Incubator kie issues#1906] New version of date and time conversion function

### LINKED ISSUE #1906: New version of date and time conversion function
In DMN 1.6, as part of the deprecation of time type with a timezone, a new conversion function is introduced: 

Image

The new version needs to be added to the existing date and time conversion function. 

Note:
please create unit-test specifically targeting, directly, the related functions

### PR DESCRIPTION
**Pull request by AthiraHari77**
Closes https://github.com/apache/incubator-kie-issues/issues/1906
This PR introduces a new conversion function, including timezone in Date and Time function.

**Thank you for submitting this pull request**

**NOTE!:** Double-check the target branch for this PR.
The default is `main` so it will target Drools 8 / Kogito.

**Ports** If a forward-port or a backport is needed, paste the forward port PR here

* [link](https://www.example.com)

**Issue**: _(please edit the GitHub Issues link if it exists)_

* [link](https://www.example.com)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* paste the link(s) from GitHub here
* link 2
* link 3 etc.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request and downstream checks</b>  
  - Push a new commit to the PR. An empty commit would be enough.

- for a <b>full downstream build</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- for <b>Jenkins PR check only</b>
  - If you are an ASF committer for KIE podling, login to Jenkins (https://ci-builds.apache.org/job/KIE/job/drools/), go to the specific PR job, and click on `Build Now` button.
</details>

### CODE REVIEW

**Comment by kie-ci3** [comment]
**PR job** `#1` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/6303 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/1/display/redirect

**Test results:**
- PASSED: 23408
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/1/testReport/org.kie.dmn.feel.runtime.functions/ComposingDifferentFunctionsTest/composite1/">org.kie.dmn.feel.runtime.functions.ComposingDifferentFunctionsTest.composite1</a></summary>
<br/>Expecting value to be true but was false
</details>

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java [hunk 1]
```diff
@@ -124,4 +127,45 @@ void invokeParamTemporalOffsetTime() {
...
+    @Test
+    void invokeParamStringDateTimeZone() {
+        FunctionTestUtil.assertResult(dateTimeFunction.invoke(LocalDate.of(2024, 12, 24),
+                        LocalTime.of(23, 59, 0),
+                        "America/Costa_Rica"),
+                ZonedDateTime.of(2024, 12, 24, 23, 59, 0, 0, ZoneId.of("America/Costa_Rica")));
+        FEELFnResult<TemporalAccessor> expectedResult = dateTimeFunction.invoke(LocalDate.of(2024, 12, 24), LocalTime.of(23, 59, 0), "America/Costa_Rica");
+        FEELFnResult<TemporalAccessor> retrievedResult = dateTimeFunction.invoke("2024-12-24T23:59:00@America/Costa_Rica");
+        assertThat(expectedResult.getOrElse(null)).isEqualTo(retrievedResult.getOrElse(null));
+    }
```
Comment (jomarko) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: apache/incubator-kie-issues#1906 has also an example where timezone is `"Z"`. Do we want to add such unit test?

**Review by jomarko — COMMENTED** [review]
Thank you for the PR

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java (hunk 1 above)
Comment (AthiraHari77) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: @jomarko I have updated the test case with the mentioned scenario where timezone is "Z"

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java [hunk 2]
```diff
@@ -124,4 +127,55 @@ void invokeParamTemporalOffsetTime() {
...
+
+    @Test
+    void invokeParamStringDateTimeZone() {
+        FunctionTestUtil.assertResult(dateTimeFunction.invoke(LocalDate.of(2024, 12, 24),
+                        LocalTime.of(23, 59, 0),
+                        "America/Costa_Rica"),
+                ZonedDateTime.of(2024, 12, 24, 23, 59, 0, 0, ZoneId.of("America/Costa_Rica")));
+        FEELFnResult<TemporalAccessor> expectedResult = dateTimeFunction.invoke(LocalDate.of(2024, 12, 24), LocalTime.of(23, 59, 0), "America/Costa_Rica");
+        FEELFnResult<TemporalAccessor> retrievedResult = dateTimeFunction.invoke("2024-12-24T23:59:00@America/Costa_Rica");
+        assertThat(expectedResult.getOrElse(null)).isEqualTo(retrievedResult.getOrElse(null));
```
Comment (jomarko) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: hmm, maybe I am missing something, here theoretically the test will pass, even if both `expectedResult` and `retrievedResult` are `null`. Am I wrong? If tno, I think we should add assertion `expectedResult` and `retrievedResult` are both not `null`.

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java [hunk 3]
```diff
@@ -124,4 +127,55 @@ void invokeParamTemporalOffsetTime() {
...
+    }
+
+    @Test
+    void testParamStringDateTimeZone() {
+        FEELFnResult<TemporalAccessor> result = dateTimeFunction.invoke(LocalDate.of(2024, 12, 24), LocalTime.of(23, 59, 0), "Z");
+        ZonedDateTime actualDateTime = (ZonedDateTime) result.getOrElse(null);
+        ZonedDateTime expectedDateTime = ZonedDateTime.of(2024, 12, 24, 23, 59, 0, 0, ZoneOffset.UTC);
+        assertThat(expectedDateTime).isEqualTo(actualDateTime);
+        FEELFnResult<TemporalAccessor> retrievedResult = dateTimeFunction.invoke("2024-12-24T23:59:00Z");
+        assertThat(actualDateTime).isEqualTo(retrievedResult.getOrElse(null));
```
Comment (jomarko) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: same comment as above, just different test.

**Review by jomarko — COMMENTED** [review]
Thank you for prompt updates on the PR. One more question put inline.

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java (hunk 2 above)
Comment (AthiraHari77) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: I have updated the tests to assert that  both expectedResult and retrievedResult are not null.

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java [hunk 4]
```diff
@@ -124,4 +127,59 @@ void invokeParamTemporalOffsetTime() {
...
     }
+
+    @Test
+    void invokeParamStringDateTimeZone() {
+        FunctionTestUtil.assertResult(dateTimeFunction.invoke(LocalDate.of(2024, 12, 24),
+                        LocalTime.of(23, 59, 0),
+                        "America/Costa_Rica"),
+                ZonedDateTime.of(2024, 12, 24, 23, 59, 0, 0, ZoneId.of("America/Costa_Rica")));
+        FEELFnResult<TemporalAccessor> expectedResult = dateTimeFunction.invoke(LocalDate.of(2024, 12, 24), LocalTime.of(23, 59, 0), "America/Costa_Rica");
+        assertThat(expectedResult).isNotNull();
```
Comment (jomarko) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: I am sorry, my point was to be sure, `expectedResult.getOrElse(null) != null`. What is slightly different from `expectedResult != null`

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java [hunk 5]
```diff
@@ -124,4 +127,59 @@ void invokeParamTemporalOffsetTime() {
...
+        FEELFnResult<TemporalAccessor> retrievedResult = dateTimeFunction.invoke("2024-12-24T23:59:00@America/Costa_Rica");
+        assertThat(retrievedResult).isNotNull();
```
Comment (jomarko) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: same about `retrievedResult`

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java [hunk 6]
```diff
@@ -124,4 +127,59 @@ void invokeParamTemporalOffsetTime() {
...
+    @Test
+    void testParamStringDateTimeZone() {
+        FEELFnResult<TemporalAccessor> result = dateTimeFunction.invoke(LocalDate.of(2024, 12, 24), LocalTime.of(23, 59, 0), "Z");
+        assertThat(result).isNotNull();
+        ZonedDateTime actualDateTime = (ZonedDateTime) result.getOrElse(null);
+        ZonedDateTime expectedDateTime = ZonedDateTime.of(2024, 12, 24, 23, 59, 0, 0, ZoneOffset.UTC);
+        assertThat(expectedDateTime).isEqualTo(actualDateTime);
+        FEELFnResult<TemporalAccessor> retrievedResult = dateTimeFunction.invoke("2024-12-24T23:59:00Z");
+        assertThat(retrievedResult).isNotNull();
+        assertThat(actualDateTime).isEqualTo(retrievedResult.getOrElse(null));
```
Comment (jomarko) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: Similar comments here

**Review by jomarko — COMMENTED** [review]
Thank you for updates. I am sorry. I was probably not clear with my point. I added more, what was in my mind about the assertion,

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 7]
```diff
@@ -58,6 +58,32 @@ public class DateAndTimeFunction
                                                                  .toFormatter();
     }
 
+    static FEELFnResult<TemporalAccessor> validateDate(TemporalAccessor date) {
```
Comment (gitgabrio) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: 👍

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 8]
```diff
@@ -58,6 +58,32 @@ public class DateAndTimeFunction
...
+            date = date.query(TemporalQueries.localDate());
+
+            if (date == null) {
+                return FEELFnResult.ofError(new InvalidParametersEvent(Severity.ERROR, "date", "must be an instance of LocalDate (or must be possible to convert to a FEEL date using built-in date(date) )"));
+            }
+        }
+        return FEELFnResult.ofResult(date);
+    }
+
+    static FEELFnResult<TemporalAccessor> validateTime(TemporalAccessor time) {
```
Comment (gitgabrio) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: 👍

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 9]
```diff
@@ -185,5 +201,30 @@ public FEELFnResult<TemporalAccessor> invoke(@ParameterName( "year" ) Number yea
...
+    public FEELFnResult<TemporalAccessor> invoke(@ParameterName("date") TemporalAccessor date, @ParameterName("time") TemporalAccessor time, @ParameterName("timeZone") String timeZone) {
+        FEELFnResult<TemporalAccessor> dateValidationResult = validateDate(date);
+        if (dateValidationResult.isLeft()) {
+            return dateValidationResult;
+        }
+        FEELFnResult<TemporalAccessor> timeValidationResult = validateTime(time);
+        if (timeValidationResult.isLeft()) {
+            return timeValidationResult;
+        }
+        if (timeZone == null || timeZone.isEmpty() || TimeZone.getTimeZone(timeZone) == null) {
```
Comment (gitgabrio) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: Since you implemented the `FEELFnResult<TemporalAccessor> validate*(TemporalAccessor)` logic, it would be worth to do the same for `timeZone` itself (validateTimeZone(timeZone) should return a result containing `ZoneId`, so its signature should be sligthly different then the others)

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java [hunk 10]
```diff
@@ -26,10 +26,13 @@
...
 import java.time.ZonedDateTime;
 import java.time.temporal.Temporal;
+import java.time.temporal.TemporalAccessor;
 
 import org.junit.jupiter.api.Test;
 import org.kie.dmn.feel.runtime.events.InvalidParametersEvent;
 
+import static org.assertj.core.api.Assertions.assertThat;
+
 class DateTimeFunctionTest {
```
Comment (gitgabrio) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateTimeFunctionTest.java]: For some reason, there are duplicated test classes for `DateAndTimeFunction`

could you please
1. move all the tests from `DateTimeFunctionTest` to `DateAndTimeFunctionTest` ? (it follow the convention {tested_clas_name}Test)
2. delete `DateTimeFunctionTest` 

Thanks! 🙏

**Review by gitgabrio — CHANGES_REQUESTED** [review]
Thanks @AthiraHari77 , good job! Just a couple of details to fix, from my POV

**Comment by gitgabrio** [comment]
@baldimir @yesamer 
Could you please take alook at it, and especially verify if the tests covers all possible use cases ? Thanks!

**Comment by AthiraHari77** [comment]
@jomarko @gitgabrio I have updated the PR with the code changes requested, please review the same

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 11]
```diff
@@ -82,30 +119,20 @@ public FEELFnResult<TemporalAccessor> invoke(@ParameterName( "from" ) String val
...
+        FEELFnResult<TemporalAccessor> timeValidationResult = validateTime(time);
+        if (timeValidationResult.isLeft()) {
+            return timeValidationResult;
         }
-
         try {
-            if( date instanceof LocalDate && time instanceof LocalTime ) {
-                return FEELFnResult.ofResult( LocalDateTime.of( (LocalDate) date, (LocalTime) time ) );
-            } else if (date instanceof LocalDate && (time.query(TemporalQueries.localTime()) != null && time.query(TemporalQueries.zone()) != null)) {
+            date = dateValidationResult.getOrElse(null);
```
Comment (gitgabrio) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: Hi @AthiraHari77 
Sorry, but that 
`dateValidationResult.getOrElse(null)`  reintroduce a potential bug that should have been take care of by 
`validateDate(date);`

1. the code correctly invoke `validateDate` , to check that it is not `null`
2. then, `dateValidationResult.getOrElse(null)` reintroduce the chance to date being `null`
3. it does not happen becuase of previous `if (dateValidationResult.isLeft())` invocation, but that `dateValidationResult.getOrElse(null)` is wrong, by itself.
If `dateValidationResult` contains `null` and not a date, for whatever reason, then an exception should be thrown.
Please replace `date = dateValidationResult.getOrElse(null);` with
`date = dateValidationResult.getOrElseThrow(_to_be_define);`

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 12]
```diff
@@ -82,30 +119,20 @@ public FEELFnResult<TemporalAccessor> invoke(@ParameterName( "from" ) String val
...
+            if (date instanceof LocalDate && time instanceof LocalTime) {
+                return FEELFnResult.ofResult( LocalDateTime.of( (LocalDate) date, (LocalTime) time ));
```
Comment (gitgabrio) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: Here, IINW, time it is still the original parameter, while `date` is the one retrieved from `dateValidationResult`:
please use a consistent way to manage equivalent variables

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 13]
```diff
@@ -185,5 +212,31 @@ public FEELFnResult<TemporalAccessor> invoke(@ParameterName( "year" ) Number yea
...
+        FEELFnResult<TemporalAccessor> timeValidationResult = validateTime(time);
+        if (timeValidationResult.isLeft()) {
+            return timeValidationResult;
+        }
+        FEELFnResult<ZoneId> timeZoneValidationResult = validateTimeZone(timeZone);
+        if (timeZoneValidationResult.isLeft()) {
+            return timeZoneValidationResult.cata(FEELFnResult::ofError, value -> null);
+        }
+        try {
+            date = dateValidationResult.getOrElse(null);
```
Comment (gitgabrio) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: @AthiraHari77 
See comments above

**Review by gitgabrio — CHANGES_REQUESTED** [review]
HI  @AthiraHari77 
Thanks, there are just few details left to fix 😄

**Comment by jomarko** [comment]
@AthiraHari77 I think I do not have more feedback than @gitgabrio provided. Thank you @gitgabrio

**Comment by kie-ci3** [comment]
**PR job** `#8` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/6303 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/8/display/redirect

**Test results:**
- PASSED: 23570
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/8/testReport/org.kie.dmn.feel.runtime.functions/TimeFunctionTest/invokeTemporalAccessorParamUnsupportedAccessor/">org.kie.dmn.feel.runtime.functions.TimeFunctionTest.invokeTemporalAccessorParamUnsupportedAccessor</a></summary>
java.lang.NullPointerException<br/>	at java.base/java.util.Objects.requireNonNull(Objects.java:209)<br/>	at java.base/java.util.Optional.of(Optional.java:113)<br/>	at org.kie.dmn.feel.runtime.functions.DateAndTimeFunction.getValidDate(DateAndTimeFunction.java:78)<br/>	at org.kie.dmn.feel.runtime.functions.DateAndTimeFunction.invoke(DateAndTimeFunction.java:123)<br/>	at org.kie.dmn.feel.runtime.functions.TimeFunction.invoke(TimeFunction.java:157)<br/>	at org.kie.dmn.feel.runtime.functions.TimeFunctionTest.invokeTemporalAccessorParamUnsupportedAccessor(TimeFunctionTest.java:98)<br/>	at java.base/java.lang.reflect.Method.invoke(Method.java:568)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>
</details>

**Comment by kie-ci3** [comment]
**PR job** `#9` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/6303 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/9/display/redirect

**Test results:**
- PASSED: 23570
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/9/testReport/org.kie.dmn.feel.runtime.functions/TimeFunctionTest/invokeTemporalAccessorParamUnsupportedAccessor/">org.kie.dmn.feel.runtime.functions.TimeFunctionTest.invokeTemporalAccessorParamUnsupportedAccessor</a></summary>
java.lang.NullPointerException<br/>	at java.base/java.util.Objects.requireNonNull(Objects.java:209)<br/>	at java.base/java.util.Optional.of(Optional.java:113)<br/>	at org.kie.dmn.feel.runtime.functions.DateAndTimeFunction.getValidDate(DateAndTimeFunction.java:78)<br/>	at org.kie.dmn.feel.runtime.functions.DateAndTimeFunction.invoke(DateAndTimeFunction.java:123)<br/>	at org.kie.dmn.feel.runtime.functions.TimeFunction.invoke(TimeFunction.java:157)<br/>	at org.kie.dmn.feel.runtime.functions.TimeFunctionTest.invokeTemporalAccessorParamUnsupportedAccessor(TimeFunctionTest.java:98)<br/>	at java.base/java.lang.reflect.Method.invoke(Method.java:568)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>
</details>

**Review by gitgabrio — APPROVED** [review]
Well done, thanks @AthiraHari77 !!

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 14]
```diff
@@ -62,10 +62,64 @@ public class DateAndTimeFunction
...
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(ZoneId.of(timeZone));
+        } catch (DateTimeException ex) {
+            return Optional.empty();
+        }
+    }
+
+    static FEELFnResult<TemporalAccessor> generateDateTimeAndTimezone(TemporalAccessor date, TemporalAccessor time, Optional<ZoneId> zoneId) {
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: niptick (not binding): Using an Optional as method parameters should be avoided. (google it for more info).  I suggest replacing it with 2 methods with different parameters list instead
eg. 
`generateDateTimeAndTimezone(TemporalAccessor date, TemporalAccessor time, Optional<ZoneId> zoneId)`
`generateDateTimeAndTimezone(TemporalAccessor date, TemporalAccessor time)`

**Comment by kie-ci3** [comment]
**PR job** `#13` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/6303 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/13/display/redirect

**Test results:**
- PASSED: 11864
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/13/testReport/org.kie.dmn.feel.runtime.functions/ComposingDifferentFunctionsTest/composite1/">org.kie.dmn.feel.runtime.functions.ComposingDifferentFunctionsTest.composite1</a></summary>
<br/>Expecting value to be true but was false
</details>

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java [hunk 15]
```diff
@@ -62,10 +62,64 @@ public class DateAndTimeFunction
...
+        Optional<TemporalAccessor> dateValidationResult = getValidDate(date);
+        Optional<TemporalAccessor> timeValidationResult = getValidTime(time);
+
+        try {
+            TemporalAccessor validatedDate = dateValidationResult.orElseThrow(() -> new NoSuchElementException("Parameter 'date' is missing or invalid."));
+            TemporalAccessor validatedTime = timeValidationResult.orElseThrow(() -> new NoSuchElementException("Parameter 'time' is missing or invalid."));
+            if (date instanceof LocalDate && time instanceof LocalTime) {
```
[the remark this answers is not in the document]
Comment (AthiraHari77) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunction.java]: Fixed

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunctionTest.java [hunk 16]
```diff
@@ -45,4 +55,219 @@ void invokeFromString() {
...
+    void invokeParamStringDateTimeZoned() {
+        FunctionTestUtil.assertResult(dateTimeFunction.invoke("2011-12-31T10:15:30@Europe/Paris"),
+                ZonedDateTime.of(2011, 12, 31, 10, 15, 30, 0, ZoneId.of("Europe/Paris")));
+        FunctionTestUtil.assertResult(dateTimeFunction.invoke("2011-12-31T10:15:30.987@Europe/Paris"),
+                ZonedDateTime.of(2011, 12, 31, 10, 15, 30, 987_000_000, ZoneId.of("Europe/Paris"
+                )));
+        FunctionTestUtil.assertResult(dateTimeFunction.invoke("2011-12-31T10:15:30.123456789@Europe/Paris"),
+                ZonedDateTime.of(2011, 12, 31, 10, 15, 30, 123_456_789, ZoneId.of("Europe/Paris"
+                )));
+        FunctionTestUtil.assertResult(dateTimeFunction.invoke("999999999-12-31T23:59:59.999999999@Europe/Paris"),
```
[the remark this answers is not in the document]
Comment (AthiraHari77) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/DateAndTimeFunctionTest.java]: Fixed

**Review by yesamer — APPROVED** [review]
The PR looks good, thank you @AthiraHari77. 
Although the behavior is correct, I'm a bit doubtful about the Optional usage we introduced in this PR. It seems to me that the Optional semantic here is to propagate an error (i.e. 
In case of wrong parameters --> 
we return empty Optional --> 
in case of empty Optional we throw an exception --> 
we catch the exception, returning the `FEELFnResult.ofError`) 
I would argue that is not the best use case to use Optional, that is more intended to manage the `value - empty` use case, and not the `value - error` case.
As an alternative solution, I would personally suggest more to directly thrown an exception in `getValidTimeZone()` `getValidDate()`. 
Approving as the above consideration is not biding, but please consider it.

**Review by jomarko — APPROVED** [review]
Thank you for the PR, especially the tests. I read them and I think we can approve this PR.

**Comment by gitgabrio** [comment]
> The PR looks good, thank you @AthiraHari77. Although the behavior is correct, I'm a bit doubtful about the Optional usage we introduced in this PR. It seems to me that the Optional semantic here is to propagate an error (i.e. In case of wrong parameters --> we return empty Optional --> in case of empty Optional we throw an exception --> we catch the exception, returning the `FEELFnResult.ofError`) I would argue that is not the best use case to use Optional, that is more intended to manage the `value - empty` use case, and not the `value - error` case. As an alternative solution, I would personally suggest more to directly thrown an exception in `getValidTimeZone()` `getValidDate()`. Approving as the above consideration is not biding, but please consider it.

Hi @yesamer - my bad: I suggested to @AthiraHari77  the usage of `Optional` but, looking at it now, it seems you are right 👍

**Comment by yesamer** [comment]
@AthiraHari77 100% approved, thank you for the changes.

**Comment by kie-ci3** [comment]
**PR job** `#15` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/6303 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/15/display/redirect

**Test results:**
- PASSED: 23577
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6303/15/testReport/org.kie.dmn.feel.runtime.functions/ComposingDifferentFunctionsTest/composite1/">org.kie.dmn.feel.runtime.functions.ComposingDifferentFunctionsTest.composite1</a></summary>
<br/>Expecting value to be true but was false
</details>

**Comment by yesamer** [comment]
I guess we can merge it, right? @gitgabrio @baldimir @AthiraHari77

**Comment by gitgabrio** [comment]
@yesamer 
Let's wait @baldimir to approve and merge

**Review by baldimir — APPROVED** [review]
